### PR TITLE
Preliminary concurrency support with mioco.

### PR DESCRIPTION
**Pull request by nhoad**
mioco runs on top of mio and uses coroutines, so they're quite comfortable to write.

However it does mean we have to run on nightly, which is a thing to consider.

Also, performance - I don't have any hard numbers, but the concurrency benchmarks given by the author are quite impressive: https://github.com/dpc/mioco/blob/master/BENCHMARKS.md

368k requests a second with httparse is pretty impressive. That's with 100 client connections as well, which would translate to 100 threads in the old version, which wasn't exactly scalable.

Before we can come up with any realistic numbers for octopus, we need to implement response parsing from upstream - at the moment it's just a dumb tcp tunnel.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,129 +1,41 @@
...
-
-        // reset the buffer so we have a clean slate for keep-alive.
-        buffer.truncate(0);
-    }
-}
-
-fn main() {
-    let port = 8000;
+            mioco::spawn(move || -> io::Result<()> {
+                handle_client(conn);
```
Comment (miselin) [inline src/main.rs]: Why not make handle_client return a Result here? Does it blow up the loop?

File: src/server.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    let resize_amount = 4096;
+
+    match len {
+        0 => {
+            println!("buffer is empty, growing {} -> {}", len, len + resize_amount);
+            buffer.resize(len + resize_amount, 0);
+        },
+        _ if len == capacity => {
+            println!("buffer is at capacity, growing {} -> {}", len, len + resize_amount);
+            buffer.resize(len + resize_amount, 0);
```
Comment (miselin) [inline src/server.rs]: While this is being changed, maybe we should consider adding a reserve() call for len \* 2 before or after the call to resize(), under the assumption that we'll most likely read more . If the function is likely to be going away, though, we can probably leave this be.

**Comment by miselin** [comment]
Looks good. A couple notes but no blockers. Let's work towards getting some benchmarks up and running - I have a hunch that it'll be easier with mioco.

File: src/main.rs (hunk 1 above)
Comment (nhoad) [inline src/main.rs]: Good point! I'll work on getting that going.

File: src/server.rs (hunk 2 above)
Comment (nhoad) [inline src/server.rs]: I started writing some tests for this and I found all sorts of awful edge cases, so I've greatly simplified it - performance may suffer a bit, but at least it's bugfree now!

**Comment by miselin** [comment]
Looks ready to submit to me. I'll look into building some benchmarks once this goes in.

**Comment by nhoad** [comment]
Awesome, cheers! I've got another change and some tests I want to get pushed a little bit later today, will poke you after that for another review.

**Comment by nhoad** [comment]
@miselin If the lifetime changes and parsing error handling looks good to you, I'd like to merge this in - I've got some other dependent stuff for doing DNS resolution and response parsing that I'd like to get started on :)

In particular returning `String` for the error types feels quite poor, but baby steps I suppose.

**Comment by miselin** [comment]
Looks good to merge so we can iterate on it further.
